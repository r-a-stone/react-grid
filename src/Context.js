import React, { createContext, useState } from "react";
import PropTypes from "prop-types";

const Context = createContext();

const propTypes = {
  columns: PropTypes.number,
  gutters: PropTypes.number,
  breakpoints: PropTypes.array
};

const defaultProps = {
  columns: 12,
  gutters: 24,
  breakpoints: [
    {
      name: "xs",
      minWidth: 0
    },
    {
      name: "sm",
      minWidth: 480
    },
    {
      name: "md",
      minWidth: 768
    },
    {
      name: "lg",
      minWidth: 1024
    }
  ]
};

const GridProvider = ({ children, ...props }) => {
  const [columns, setColumns] = useState(props.columns);
  const [gutters, setGutters] = useState(props.gutters);
  const [breakpoints, setBreakpoints] = useState(props.breakpoints);

  return (
    <Context.Provider
      value={{
        columns,
        gutters,
        breakpoints,

        setColumns,
        setGutters,
        setBreakpoints
      }}
    >
      {children}
    </Context.Provider>
  );
};

GridProvider.propTypes = propTypes;
GridProvider.defaultProps = defaultProps;

export default {
  Provider: GridProvider,
  Consumer: Context.Consumer
};
