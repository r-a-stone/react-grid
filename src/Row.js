import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

const propTypes = {
  flexDirection: PropTypes.oneOf([
    "row",
    "row-reverse",
    "column",
    "column-reverse"
  ]),
  flexWrap: PropTypes.oneOf(["wrap-reverse", "wrap", "nowrap"]),
  justifyContent: PropTypes.oneOf([
    "flex-start",
    "center",
    "flex-end",
    "space-between",
    "stretch",
    "normal"
  ]),
  alignItems: PropTypes.oneOf([
    "flex-start",
    "center",
    "flex-end",
    "baseline",
    "stretch",
    "normal"
  ]),
  alignContent: PropTypes.oneOf([
    "flex-start",
    "center",
    "flex-end",
    "baseline",
    "stretch",
    "normal"
  ])
};

const defaultProps = {
  flexDirection: "row",
  flexWrap: "wrap",
  justifyContent: "normal",
  alignItems: "normal",
  alignContent: "normal"
};

export const StyledRow = styled("div")`
  box-sizing: border-box;
  &:before,
  &:after {
    box-sizing: border-box;
  }

  margin-right: ${props => `${props.theme.gutters / 2 / 16}rem`};
  margin-left: ${props => `${props.theme.gutters / 2 / 16}rem`};

  display: flex;
  flex-direction: ${props => props.flexDirection};
  flex-wrap: ${props => props.flexWrap};
  justify-content: ${props => props.justifyContent};
  align-items: ${props => props.alignItems};
  align-content: ${props => props.alignContent};
`;

const Row = ({
  children,
  flexDirection,
  flexWrap,
  justifyContent,
  alignItems,
  alignContent,
  gutters,
  ...props
}) => {
  return (
    <StyledRow
      {...props}
      theme={{
        gutters,
        flexDirection,
        flexWrap,
        justifyContent,
        alignItems,
        alignContent
      }}
    >
      {children}
    </StyledRow>
  );
};

Row.propTypes = propTypes;
Row.defaultProps = defaultProps;

export default Row;
