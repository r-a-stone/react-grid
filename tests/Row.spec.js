import React from "react";
import renderer from "react-test-renderer";
import Context from "../src/Context";
import Row from "../src/Row";

it("Row renders without crashing.", () => {
  const component = renderer.create(<Row gutters={24}>Row Content</Row>);

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it("Row with context renders without crashing.", () => {
  const component = renderer.create(
    <Context.Provider>
      <Context.Consumer>
        {({ gutters }) => <Row gutters={gutters}>Row Content</Row>}
      </Context.Consumer>
    </Context.Provider>
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
