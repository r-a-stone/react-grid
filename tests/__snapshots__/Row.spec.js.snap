// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Row renders without crashing. 1`] = `
<div
  className="sc-bdVaJa ewlvnh"
>
  Row Content
</div>
`;

exports[`Row with context renders without crashing. 1`] = `
<div
  className="sc-bdVaJa ewlvnh"
>
  Row Content
</div>
`;
